### DISCUSSION TITLE
Setting a counter in STM32

### DISCUSSION THREAD

**Discussion by ordonezt** [zephyrproject-rtos/zephyr#47033] [Q&A] [answered]
Hi im trying to use the counter API with stm32_min_dev_blue board.
This is my app.overlay:

```
&timers2 {
	status = "okay";
	counter {
		compatible = "st,stm32-counter";
		status = "okay";
		label = "COUNTER_2";
	};
};
```

And this is the zephyr.dts file after the build

```
timers2: timers@40000000 {
			compatible = "st,stm32-timers";
			reg = < 0x40000000 0x400 >;
			clocks = < &rcc 0x1c 0x1 >;
			interrupts = < 0x1c 0x0 >;
			interrupt-names = "global";
			st,prescaler = < 0x0 >;
			status = "okay";
			label = "TIMERS_2";
			pwm {
				compatible = "st,stm32-pwm";
				status = "disabled";
				label = "PWM_2";
				#pwm-cells = < 0x3 >;
			};
			counter {
				compatible = "st,stm32-counter";
				status = "okay";
				label = "COUNTER_2";
			};
		};
```

This is my prj.conf

```
CONFIG_SERIAL=y
CONFIG_GPIO=y
CONFIG_PRINTK=y
CONFIG_PWM=y

CONFIG_COUNTER=y
CONFIG_COUNTER_INIT_PRIORITY=50
CONFIG_COUNTER_TIMER_STM32=y
CONFIG_COUNTER_RTC_STM32=n
```

This is my code

```
#include <zephyr.h>
#include <zephyr/drivers/counter.h>

/* Counter */
struct counter_alarm_cfg alarm_cfg;
#define TIMER DT_NODELABEL(timers2)
const struct device *counter_dev = DEVICE_DT_GET(TIMER);

#if !DT_NODE_EXISTS(TIMER)
#error "whoops"
#endif
```

If i build this everything is fine.
Then if i add this code the build fails in this line `const struct device *counter_dev = DEVICE_DT_GET(TIMER);` with the following output

```
/* Threads */
#define GEN_STACK_SIZE   500
#define GEN_PRIORITY     5
void gen_thread(void *, void *, void *);

K_THREAD_DEFINE(gen_tid, GEN_STACK_SIZE,
                gen_thread, NULL, NULL, NULL,
                GEN_PRIORITY, 0, 0);

void gen_thread(void *arg1, void *arg2, void *arg3){
    ARG_UNUSED(arg1);
    ARG_UNUSED(arg2);
    ARG_UNUSED(arg3);

    if (!device_is_ready(counter_dev)) {
    printk("Error: Counter device %s is not ready\n",
          counter_dev->name);
      return;
    }

    while (1) {
		printk("Thread gen loopeando\n");
		k_sleep(K_MSEC(50));
	}
}
```
```
c:/users/tomaso/zephyr-sdk-0.14.2/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/10.3.0/../../../../arm-zephyr-eabi/bin/ld.exe: app/libapp.a(gen.c.obj):C:/Users/tomaso/zephyrproject/proyectos/pruebaaa/src/gen.c:19: undefined reference to '__device_dts_ord_63'
collect2.exe: error: ld returned 1 exit status
ninja: build stopped: subcommand failed.
FATAL ERROR: command exited with status 1: 'C:\Program Files\CMake\bin\cmake.EXE' --build 'C:\Users\tomaso\zephyrproject\build'
```

What is the correct way to include a counter to a stm32 timer? I'm missing something in the overlay?

**Comment by erwango** [comment]
`CONFIG_PWM=y` likely conflicts with the fact that no pwm node is enabled:
```
			pwm {
				compatible = "st,stm32-pwm";
				status = "disabled";
```

**Reply by ordonezt** [reply]
Hi Erwang, i changed the overlay and binding to the following and the project started to build (im quite new to zephyr and dt :sweat_smile:), i will test it later on a board and let you know the results.

app.overlay:
```
&timers2 {
	status = "okay";
	counters2: counter {
		compatible = "st,stm32-counter";
		status = "okay";
		label = "COUNTER_2";
	};
};
```

binding:
```
#define TIMER DT_NODELABEL(counters2)
const struct device *counter_dev = DEVICE_DT_GET(TIMER);
```

**Comment by ordonezt** [comment, marked as answer]
Ok, it worked, this is the solution https://github.com/zephyrproject-rtos/zephyr/discussions/47033#discussioncomment-3065451

**Reply by mrcamilletti** [reply]
Hi @ordonezt ! I'm currently working with the same board and I also want to setup a timer (counter). I followed your approach and tried using counter_set_top_value and counter_set_channel_alarm (one or other), however after calling counter_start() the callbacks were never called. Is it there a an aditional configuration required? did you have to setup the prescaler?

**Reply by ordonezt** [reply]
Hi @mrcamilletti! It's been a long time since I last opened this project, but from what I remember I chose to use a software timer instead of a hardware one. I remember at that time there was no support for stm32f1 counters or something like that, maybe that has changed.

**Reply by mrcamilletti** [reply]
Ok I'll try a bit more and If doesn't work then I will open a new issue thanks!

**Reply by ordonezt** [reply]
Good luck!

**Reply by mrcamilletti** [reply]
Solved. I was using timer1 instead of timer2... it seems counter does not work with timer1. I believe it has something to do with the interrupts configuration.

**Reply by athanPrio** [reply]
Hi @mrcamilletti ! I appreciate if you could write a sample of your code about setting up a stm32 timer as a counter.

**Comment by tguenth** [comment]
I came a cross the same issue on a nucleo_g474re board with the timer to not started as expected with timer shell. I propose to check  the timer enable flag if it actually got set. Thus never triggered an interrupt handler.

shell gets stuck on both 
```shell
uart:~$ timer periodic COUNTER_2 1000
uart:~$ timer oneshot COUNTER_2 0 1000
```

with the debugger realized the timer got never enabled. maybe this the same in your case.

![image](https://github.com/user-attachments/assets/818b26b5-a1e3-4f58-8ad5-d6510ae0c6b7)

manually writing a `1` into CEN (counter enable) will make the shell command return and you will have a responsive shell again.

```shell
uart:~$ timer periodic COUNTER_2 1000
COUNTER_2: periodic timer triggered for 10 times
``` 
I'm not sure if it is intended behavior to have the shell get stuck on k_sem_take with the semaphore never being set, as the timer_alarm_handler never gets called. From my understanding no shell command for counters actually enables them. From the snipped below you see the counter to be configured at `counter_set_channel_alarm` an then directly awaiting a singal from an ISR handler.

(https://github.com/zephyrproject-rtos/zephyr/blob/495c3b4d26f1b610ce63059430eab0b5a0842c40/drivers/counter/counter_timer_shell.c).
``` c
  	/* set an alarm */
	err = counter_set_channel_alarm(timer_dev, (uint8_t)channel, &alarm_cfg);
	if (err != 0) {
		shell_error(shctx, "%s:Failed to set channel alarm, err:%d", argv[ARGV_DEV], err);
		return err;
	}

	k_sem_take(&timer_sem, K_FOREVER); # never gets passed until counter is acutally enabled

	shell_info(shctx, "%s: Alarm triggered", argv[ARGV_DEV]);

	return 0;
}
```

my sources to reproduce that

main.c
``` c
#include <zephyr/kernel.h>

int main(void)
{
	while (1) {
		k_sleep(K_FOREVER);
	}
	return 0;
}
```

proj.conf
```conf
CONFIG_SHELL=y
CONFIG_COUNTER=y
CONFIG_COUNTER_SHELL=y
```

nucleo_g474re.overlay
```txt
&timers2 {
	st,prescaler = <79>;
	counter {
		status = "okay";
		label = "counter_2";
	};
};
```
